{% extends 'base.html' %}
{% block pagetitle %}Getting Started{% endblock %}
{% block css%}
.jumbotron {
background: -webkit-linear-gradient(#2E82D1, #7CC1EC); /* For Safari */
background: -o-linear-gradient(#2E82D1, #7CC1EC); /* For Opera 11.1 to 12.0 */
background: -moz-linear-gradient(#2E82D1, #7CC1EC); /* For Firefox 3.6 to 15 */
background: linear-gradient(#2E82D1, #7CC1EC); /* Standard syntax (must be last) */	
}

.e{
	background-color:#FFFFFF;
}
.e:hover{
   //background-color:#99CCFF;
}

h1{ position:relative;}
h1 span{ position:absolute; top:-70px;}

h2{ position:relative;}
h2 span{ position:absolute; top:-70px;}

h3{ position:relative; }
h3 span{ position:absolute; top:-70px;}

{% endblock %}

{% block javascript%}

<script type="text/javascript">
	$(window).resize(function() {
	  if($(window).width() < 753){
		$('#navbar').hide();
	  }else{ 
		$('#navbar').show();}
	});

   $(window).scroll(function() {
	
		if ($(this).scrollTop()>270 )
		 {
		    $('#navbar').css({position:"fixed",top: "80px"});
			
		 }
		else
		 {
			$('#navbar').css({position:"absolute",top: "30px"});
		 }
	 });

</script>

{% endblock %}

{% block header %}
<!-- Main jumbotron for a primary marketing message or call to action -->
    <div class="jumbotron">
      <div class="container">
        <h1 style ="color:#fff">Getting Started</h1>
        <h3 style = "color:#E3FBFF">A quick guide to use our website</h3>
      </div>
    </div>

{% endblock %}    
<!-- Contents -->




{% block content %}
<div class="row row-offcanvas row-offcanvas-right"> 
	<div class="col-xs-12 col-sm-9"><!-- Left Canvas-->


<!--########################### Image Modal ############################-->
				<script>
				$(document).ready(function(){
					$("img").click(function(){
						var Url = $(this).data('ref');
						$("#imgarea").html("<img src='" + Url + "' class='img-responsive' >" );
						$("#myModalLabel").html($(this).data('head'));
					});
				});
				</script>
				<!-- Modal -->
				<div class="modal fade" id="myModalimg" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
				  <div class="modal-dialog modal-lg">
					<div class="modal-content">
					  <div class="modal-header">
						<button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>      
						<h4 class="modal-title" id="myModalLabel"></h4>
					  </div>
					  <div class="modal-body">
						<div id="imgarea" style="text-align: center;"></div>
					  </div>
					  <div class="modal-footer">
						<button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
					  </div>
					</div>
				  </div>
				</div><!-- END Modal -->


<!--########################### Overview ############################-->
	<h1 class="page-header"><span id="basic"></span>3 Steps to View</h1>
 <p class="lead">Viewing information can be very easy within only 3 simple steps!</p>
		<h3>Search</h3>
		<p>The first step is to retrieve what you are actually interested in from all the information stored in the database. Depending on your needs, SNVdb offers three search modes: by Uniprot accession number, by SNV FTID or by disease name.</p>

		<p>Just select your preferred search mode from the drop-down menu in the home page, and enter your query in the search bar below.</p>


		<h3>Select</h3>


		<p>If your search returns more than one entry from the database, you will be taken to the Search Results page. Simply click on your favourite entry from the list.</p>

		<p>In case you mistyped something, you will be able to see the search term that you used and type the right word in.</p>

		<h3>Visualise</h3>

		<p>Finally, you can see your entry here. Depending on whether it is a protein, a SNV, a disease, an interaction or an interface, the page will look slightly different:</p>

			<div class="panel panel-default">
			  <div class="panel-body">
				
				  <!-- columns -->
				  <div class="row">
					<div class="col-md-4 e">
					  <img src="{{STATIC_URL}}search.jpeg"  class="img-thumbnail" data-toggle="modal" data-target="#myModalimg" data-head="Search" data-ref= "{{STATIC_URL}}search-l.jpg" alt="image">
					  <p class="text-center">Search</p>
					</div>
					<div class="col-md-4 e">
					  <img src="{{STATIC_URL}}select.jpeg"  class="img-thumbnail" data-toggle="modal" data-target="#myModalimg" data-head="Select" data-ref= "{{STATIC_URL}}select-l.jpg" alt="image">
					  <p class="text-center">Select</p>
				   </div>
					<div class="col-md-4 e">
					  <img src="{{STATIC_URL}}view.jpeg"  class="img-thumbnail"  data-toggle="modal" data-target="#myModalimg" data-head="Visualise" data-ref= "{{STATIC_URL}}view-l.jpg"  alt="image">
					  <p class="text-center">Visualise</p>
					</div>
				  </div>
			  </div>
			</div>


<!--########################### Uniprot ############################-->

		<h1 class="page-header"><span id="uniprot"></span>Understanding Uniprot entries</h1>

			<p>An Uniprot entry describes a protein in the database. It is defined by the ID number by which you can find the protein in Uniprot.</p>
		


			<p>These entries are divided in five sections:</p>
			<ul>
				<li><a href="#uniprot-basic">Basic information</a></li>
				<li><a href="#uniprot-overall">Overall graphics</a></li>
				<li><a href="#uniprot-annotations">Annotations</a></li>
				<li><a href="#uniprot-interactions">Interactions</a></li>
				<li><a href="#uniprot-snvs">SNVs</a></li>
			</ul>

			<h3><span id="uniprot-basic"></span>Basic information</h3>

			<p>Here you can find some useful data about the protein and the gene encoding for it (including links to NCBI/Gene and Uniprot) as well as the link to the raw sequence (with an option to export it as a FASTA file).</p>
			<p>If the protein has any associated interactions, you can also find an image of the 3D structure of the protein.</p>
			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}basic2.jpg" class="img-responsive" alt="image">
			  </div>
			</div>	

			<p>Also, you can toggle each of the other sections on and off by clicking on the tick boxes in the "Display" box.</p>
			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}tickbox.jpg" class="img-responsive" alt="image">
			  </div>
			</div>


			<h3><span id="uniprot-overall"></span>Overall graphics</h3>

			<p>Whilst the SNVs map to defined sequence positions for a given protein, the interface residues for each of its interacting partners vary.</p>
			<p>In this section, you can see both things at the same time on a graph. The "Default" tab shows only the SNVs.</p>
			<p>The top, blue rectangle represents the protein sequence, with the SNVs plotted in red, green, blue or purple (depending on their association with disease).</p>
			<p>The grey rectangle shows the Pfam domains detected in the protein sequence, highlighting them in red and green, alternatively. Hover over each domain to see what kind of domain it is, along with some data associated with it.</p>
			<p>To add the information about interface residues in the interaction with a particular partner, just go to the top of this section and click on the corresponding tab. The interface residues of the protein for a that partner will be highlighted in yellow.</p>
			<p>Finally, hover over a highlighted residue to see some more information about it. If you click on a residue with a SNV, you will go to the SNV entry describing it.</p>
			<p>All this information is also plotted residue by residue in the detailed alignment below.</p>
			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}overall.jpg" class="img-responsive" alt="image">
			  </div>
			</div>		
	
			<h3><span id="uniprot-annotations"></span>Pfam domains</h3>

			<p>This section describes the Pfam domains that have been detected for the protein. The domain positions are plotted with respect to the protein sequence, and you can also see a list with the positions where the sequence matches the Pfam HMM for that domain, along with the statistical parameters that validate the domain.</p>
			<p>If you want to find out more about that domain, and other proteins sharing the same domain, you can click on the external link to Pfam, where you will find more information about domains.</p>

			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}pfam.jpg" class="img-responsive" alt="image">
			  </div>
			</div>


			<h3><span id="uniprot-interactions"></span>Interactions</h3>

			<p>Here we include a <strong>list of binary interactions for the protein</strong> of interest. For each one, you can see its <strong>ID on our database</strong>, the two proteins taking part in the interaction, and whether the structure of the complex has been experimentally determined or computationally modelled.</p>
			<p>Under this information, the site displays a representation of the two proteins, with the interface residues marked in yellow.</p>
			<p>There are also buttons that will lead you to the interaction entry (where you can see the 3D structure of the complex via JsMOL), the interface entry (where you can see the atom-atom interactions that hold the complex together) and the PDB file describing it.</p>
			<p>If you prefer to have the PDB file to visualize it your preferred viewer (such as PyMOL), you can also download it from here.</p>
			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}inter.jpg" class="img-responsive" alt="image">
			  </div>
			</div>

			<h3><span id="uniprot-snvs"></span>SNVs</h3>

			<p>Here you can find a list of SNVs that have an effect on the protein sequence. For each one, we show the position, the change on the sequence, the FT id and the rs id (so that you can find it in Uniprot and dbSNP).</p>
			<p>If the SNV is associated with a disease, you can also see which disease it is, and the OMIM code describing it.</p>
			<p>To find out more about each SNV, just click on the FT id link and we will take you to the <a href="#snv">SNV entry</a> describing it on our database. We also provide some links to external databases (OMIM and dbSNP). If so many links seem a bit confusing, just hover on each link to see if it links to one of the entries in SNVdb, or it leads to an external database.</p>

			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}snvs.jpeg" class="img-responsive" alt="image">
			  </div>
			</div>

<!--########################### SNVS ############################-->

		<h1 class="page-header"><span id="snv"></span>Understanding SNV entries</h1>

			<p>A SNV entry describes a single nucleotide variation affecting a particular protein. These entries are divided in <strong>four sections</strong>:</p>

			<ul>
				<li><a href="#snv-basic">Basic information</a></li>
				<li><a href="#snv-disorders">Disorders associated to the SNV</a></li>
				<li><a href="#snv-mapping">Mapping mutated SNV to the sequence</a></li>
				<li><a href="#snv-others">Other SNVs sharing the same position</a></li>
			</ul>

			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}snv_view.jpg" class="img-responsive" alt="image">
			  </div>
			</div>	


			<h3><span id="snv-basic"></span>Basic information</h3>

			<p>Here we provide some data about the SNV, such as the <strong>protein and position</strong> that it maps to (according to the Uniprot numbering), the <strong>change in sequence</strong> that it causes, and whether it is <strong>associated to any disease</strong>. If the SNV has a <strong>dbSNP id</strong> (rs ID), you can also see it here, and find a <strong>link to the dbSNP entry</strong>.</p>

			<h3><span id="snv-disorders"></span>Disorders associated to the SNV</h3>

			<p>This is a list of the <strong>diseases directly associated</strong> to this variation. You can find the <strong>link to the <a href="#disease">SNVdb disease entry</a></strong> and the <strong>OMIM code and link</strong> for each disease.</p>
		
			<h3><span id="snv-mapping"></span>Protein sequence context of the SNV:</h3>

			<p>Here you can see the SNV mapped to the <strong>sequence of the protein</strong> that it is associated to. The residue corresponding to the SNV is highlighted.</p>

			<h3><span id="snv-others"></span>Other SNVs sharing the same position</h3>

			<p>If there are <strong>more SNVs mapping to this position in the protein</strong>, you can find a list here, along with some data and the <strong>link to the SNVdb entry</strong> for each SNV.</p>

<!--########################### Disease ############################-->

		<h1 class="page-header"><span id="disease"></span>Understanding disease entries</h1>

			<p>A disease entry displays information about a particular disorder. These entries are divided in four sections:</p>

			<ul>
				<li><a href="#disease-basic">Basic information</a></li>
				<li><a href="#disease-related-diseases">Related diseases summary</a></li>
				<li><a href="#disease-related-snvs">Related SNVs list</a></li>
				<li><a href="#disease-related-uniprots">Related Uniprots</a></li>
			</ul>

			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}disease.jpeg" class="img-responsive" alt="image">
			  </div>
			</div>

			<h3><span id="disease-basic"></span>Basic information</h3>

			<p>Here we display the disease ID (which is the same as the OMIM ID) and the name of the disease.</p>

			<h3><span id="disease-related-diseases"></span>Related diseases summary</h3>

			<p>Here you can find diseases sharing some SNVs with the disease of interest. Just click on the links and see the disease entries for them.</p>

			<h3><span id="disease-related-snvs"></span>Related SNVs list</h3>

			<p>This is a list of the SNVs associated with this disease. Click on the link on the "ID" column and see the SNV entries for them.</p>

			<h3><span id="disease-related-uniprots"></span>Related proteins and their associated diseases</h3>

			<p>Some proteins are associated with more than one disease. If any of the proteins associated with the disease of interest are associated with other disorders, you will see them in this list.</p>

<!--########################### INTERFACE ############################-->

		<h1 class="page-header"><span id="interface"></span>Understanding interface entries</h1>
			<p><i>IN DEVELOPMENT</i></p>



<!--########################### INTERACTION ############################-->			

		<h1 class="page-header"><span id="interaction"></span>Understanding interaction entries</h1>
			<p>An interaction entry displays information about the three-dimensional structure of a complex between two proteins (named <strong>"partners"</strong>), capable of interacting with each other.</p>

			<div class="panel panel-default">
			  <div class="panel-body">
				<img src="{{STATIC_URL}}interaction.jpg" class="img-responsive" alt="image">
			  </div>
			</div>

			<p>These entries are divided in three parts:</p>
			<ul>
				<li>A <strong>header</strong>, displaying the Uniprot ID of the two partners, and the internal ID of the interaction in our database.</li>
				<li>A <strong>summary</strong>, displaying general information about the interaction:</li>
				<ul>
					<li>The <strong>type</strong> reflects whether the interaction structure has been experimentally determined (<i>Structure</i>), computationally modelled from another complex whose partners have high sequence identity with the partners shown (<i>Model</i>), or computationally modelled knowing that two domains detected in the partners have been shown to interact in homologous proteins (<i>Dom-dom model</i>).</li>
					<li>The <strong>names and Uniprot IDs</strong> of the two partners make it possible to find their Uniprot entries on the database. The <strong>chain IDs</strong> are also included so that you can have an idea of how well the protein is represented on this structure.</li>
					<li>The <strong>filename</strong> is the internal name of the PDB file representing the interaction, which you can also download from the Uniprot entry of any of the partners.</li>
				</ul>
				<li>The <strong>interaction structure viewer</strong>, where you can see the 3D structure of the complex, and map different features of the partners on it (see below)</li>
			</ul>


			<h3>Using the structure viewer and the feature mapping interface</h3>

			<p>The interaction PDB file is displayed on a <strong>JSMol plugin</strong>. If you already know how to use this molecule viewer, you will be happy to know that all the features of JSMol are available on the plugin. Just right-click on the structure to see the available commands on JSMol GUI, including a console to use command-line instructions).</p>

			<p>If this is your first time with JSMol, here are <strong>three basic actions</strong> that will help you explore the structure of the complex:</p>
			<ul>
				<li><strong>Hover</strong> on an atom of the structure to see the amino acid, position, chain and atom.</li>
				<li><strong>Click and drag</strong> the mouse to rotate the structure.</li>
				<li><strong>Scroll your mouse wheel</strong> up or down to zoom in and out the structure.</li>
			</ul> 

			<p>However, we have set up some <strong>quick buttons</strong> to allow you to map the features stored in the database onto the structure. You can find them right next to the plugin, and they are divided in <strong>four tabs: Basic, SNV, PFAM and Contact</strong>. Just click on each one to access its features.</p>

			<h4>The Basic tab</h4>

			<div class="row">
				<div class="col-md-6">
					<p>This tab includes some general options about the structure. Here you can:</p>

					<ul>
						<li>Choose between a <strong>cartoon</strong> representation (to have a general view of the secondary structure) or a <strong>wireframe</strong> representation (to get a detailed view of every atom in the complex).</li>
						<li><strong>Hide or show each partner</strong>, to get a clear view of the interface residues of the other partner.</li>
						<li><strong>Highlight the interface residues</strong>, which will be displayed as a <strong>spacefill</strong> model.</li>
						<li><strong>Reset everything</strong>, and go back to the initial state of the structure.</li>
					</ul>
				</div>

				<div class="col-md-6">
					<img align="center" src="{{STATIC_URL}}interaction-basic.jpeg"  class="img-thumbnail" data-toggle="modal" data-target="#myModalimg" data-head="The Basic tab" data-ref= "{{STATIC_URL}}interaction-basic.jpeg" alt="image">
						<p class="text-center">The Basic tab</p>
				</div>
			</div>
			 
    
        	<h4>The SNV tab</h4>

        	<div class="row">
        		<div class="col-md-6">
        			<p>This tab includes the options to map SNVs onto the 3D structure. Here you can:</p>
					<ul>
						<li><strong>Map or unmap all the SNVs</strong>.</li>
						<li><strong>Map or unmap a particular SNV.</strong>They are listed by partner, and sorted by the position on the PDB file of the interaction (so that the first residue of partner 2 is never #1, but the number after the position for the last residue of partner 1).</li>
					</ul>
		        
					<p>The <strong>SNVs color</strong> will be different according to their disease association:</p>
					<ul>
						<li>SNVs with at least one associated <strong>disease</strong> (according to Uniprot) will be mapped in <strong>red</strong>.</li>
						<li>SNVs classsified as <strong>polymorphisms</strong> (those not included in OMIM) will be mapped in <strong>green</strong>.</li>
						<li><strong>Unclassified SNVs (those detected in a pathological sample, but without statistically conclusive results to establish an association) </strong> will be mapped in <strong>blue</strong>.</li>
					</ul>
        		</div>

        		<div class="col-md-6">
        			<img align="center" src="{{STATIC_URL}}interaction-snv.jpeg"  class="img-thumbnail" data-toggle="modal" data-target="#myModalimg" data-head="The SNV tab" data-ref= "{{STATIC_URL}}interaction-snv.jpeg" alt="image">
        				<p class="text-center">The SNV tab</p>
        		</div>
        	</div>

			<h4>The PFAM tab</h4>

			<div class="row">
        		<div class="col-md-6">
        			<p>This tab includes the options to map Pfam domains and families onto the 3D structure. Here you can:</p>
					<ul>
						<li><strong>Map or unmap all the Pfam domains</strong> for both partners.</li>
						<li><strong>Map or unmap a particular domain</strong>. Again, they are listed by partner, and sorted by the first residue according to the PDB file.</li>
					</ul>

					<p>The different Pfam hits will be mapped in different <strong>colors</strong>:</p>
					<ul>
						<li><strong>Domains</strong> will be mapped in <strong>green</strong>.</li>
						<li><strong>Families</strong> will be mapped in <strong>pink</strong>.</li>
						<li><strong>Repeats</strong> will be mapped in <strong>range</strong>.</li>
					</ul>
        		</div>

        		<div class="col-md-6">
        			<img align="center" src="{{STATIC_URL}}interaction-pfam.jpeg"  class="img-thumbnail" data-toggle="modal" data-target="#myModalimg" data-head="The PFAM tab" data-ref= "{{STATIC_URL}}interaction-pfam.jpeg" alt="image">
        				<p class="text-center">The Pfam tab</p>
        		</div>
        	</div>

		
			<h4>The contacts tab</h4>

			<div class="row">
        		<div class="col-md-6">
        			<p>This tab includes all the options to map the bonds that keep the complex together. Here you can:</p>
					<ul>
						<li><strong>Map or unmap all</strong> the bonds.</li>
						<li>Map or unmap all the bonds of a <strong>certain type</strong>:</li>
						<ul>
							<li><strong>Disulfide bonds</strong> (in <strong>red</strong>).</li>
							<li><strong>Hydrogen bonds</strong> (in <strong>blue</strong>).</li>
							<li><strong>Polar interactions</strong> (in <strong>green</strong>).</li>
							<li><strong>Hydrophobic interactions</strong> (in <strong>yellow</strong>).</li>
							<li><strong>Unknown interactions</strong> (in <strong>purple</strong>).</li>
						</ul>
					</ul>
        		</div>

        		<div class="col-md-6">
        			<img align="center" src="{{STATIC_URL}}interaction-contacts.jpeg"  class="img-thumbnail" data-toggle="modal" data-target="#myModalimg" data-head="The Contacts tab" data-ref= "{{STATIC_URL}}interaction-contacts.jpeg" alt="image">
        				<p class="text-center">The Contacts tab</p>
        		</div>
        	</div>

			<div class="alert alert-warning">
				<h3>Note on the contact mapping</h3>

				<p><i>Please note that, to simplify the interaction and minimise loading times, <strong>only one bond per residue pair is displayed</strong>, and <strong>always between the alpha carbons of both residues</strong> (even though the atoms actually establishing the interaction might be others in the same residues).</i></p>

				<p><i>The priority to choose from the different bond types for each residue pair is:</i></p>

				<p align="center"><i><strong>Disulfide bond > Hydrogen bond > Polar interaction > Hydrophobic interaction > Unknown interaction</strong></i></p>

				<p><i>This means, for example, that hydrogen bonds between two residues that also share a disulfide bond, or hydrophobic interactions between two residues that also establish polar interactions with each other, will not be displayed.</i></p>

			</div>

<!--########################### SUPERPOSED ############################-->

<h1 class="page-header"><span id="superposed"></span>Understanding superposed structure entries</h1>
	<p><i>IN DEVELOPMENT</i></p>

<!--########################### COMPATIBILITY ############################-->
<div>
<h1 class="page-header"><span id="browser"></span>Browser and Device</h1>
  <p class="lead">This website is based on <a href = "http://getbootstrap.com" target="_blank">Bootstrap</a> framework which supports several desktop and mobile browsers. However, old browsers might not display our website correctly.</p>
<div class="alert alert-info">
  <h3>Recommendation</h3>
	<p>Although our website can run on mobile devices, it is reccommended that you browse our website using <strong>latest versions of web browsers</strong> shown above from your <strong>desktop PC</strong> with screen resolution <strong>not less than 970 px</strong> for the best display.</p>
</div>
  
  <h3 id="support-browsers">Supported browsers</h3>
  <p>Supported browsers and devices are shown below. This table is taken from the official <a href = "http://getbootstrap.com/getting-started/#support" target="_blank">Bootstrap</a> website.</p>
  <div class="table-responsive">
    <table class="table table-bordered table-striped">
      <thead>
        <tr>
          <td></td>
          <th>Chrome</th>
          <th>Firefox</th>
          <th>Internet Explorer</th>
          <th>Opera</th>
          <th>Safari</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th>Android</th>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-danger"><span class="glyphicon glyphicon-remove"></span> <span class="sr-only">Not Supported</span></td>
          <td class="text-muted" rowspan="3" style="vertical-align: middle;">N/A</td>
          <td class="text-danger"><span class="glyphicon glyphicon-remove"></span> <span class="sr-only">Not Supported</span></td>
          <td class="text-muted">N/A</td>
        </tr>
        <tr>
          <th>iOS</th>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-muted">N/A</td>
          <td class="text-danger"><span class="glyphicon glyphicon-remove"></span> <span class="sr-only">Not Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
        </tr>
        <tr>
          <th>Mac OS X</th>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
        </tr>
        <tr>
          <th>Windows</th>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-success"><span class="glyphicon glyphicon-ok"></span> <span class="sr-only">Supported</span></td>
          <td class="text-danger"><span class="glyphicon glyphicon-remove"></span> <span class="sr-only">Not Supported</span></td>
        </tr>
      </tbody>
    </table>
  </div>
  <p>Note that Internet Explorer 6 or lower are not supported. For Linux platform, the website should look and behave well enough in Chrome and Firefox.</p>


</div>



	</div><!--/ Left Canvas COL span-->	
<!-- ################## NAVIGATION BAR ##################################-->
	<div class="col-xs-6 col-sm-3 sidebar-offcanvas" id="sidebar" role="navigation">
		<div id ="navbar" class="span3 bs-docs-sidebar">
			<ul class="nav nav-list bs-sidenav">
				<li><strong><span class="glyphicon glyphicon-search"></span> Quick Navigation</strong></li>
				<li><a href="#basic"><span class="glyphicon glyphicon-chevron-right"></span> 3 Steps to View</a></li>
				<li><a href="#uniprot"><span class="glyphicon glyphicon-chevron-right"></span> Uniprot Entries</a></li>
				<li><a href="#snv"><span class="glyphicon glyphicon-chevron-right"></span> SNV Entries</a></li>
				<li><a href="#disease"><span class="glyphicon glyphicon-chevron-right"></span> Disease Entries</a></li>
				<li><a href="#interface"><span class="glyphicon glyphicon-chevron-right"></span> Interface Entries</a></li>
				<li><a href="#interaction"><span class="glyphicon glyphicon-chevron-right"></span> Interaction Entries</a></li>
				<li><a href="#superposed"><span class="glyphicon glyphicon-chevron-right"></span> Superposed Structure Entries</a></li>
				<li><a href="#browser"><span class="glyphicon glyphicon-chevron-right"></span> Browser and Device</a></li>

			</ul>
		</div><!--/span-->
	</div><!--/col-->


</div><!--/row-->


{% endblock %}

